Improve harness generated code readability

xbauch opened the pull request:
Previously all recursive initialisation took place in a single function. Now each type has it's own constructor function, which are called recursively. The results (variables to be initialised) are passed by pointer.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -195,6 +197,9 @@ void function_call_harness_generatort::implt::generate(
 
   generate_nondet_globals(function_body);
   call_function(arguments, function_body);
+  for(const auto &global_pointer : global_pointers)
```
:pick: braces please

xbauch replied:
Added.

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -195,6 +197,9 @@ void function_call_harness_generatort::implt::generate(
   generate_nondet_globals(function_body);
   call_function(arguments, function_body);
+  for(const auto &global_pointer : global_pointers)
+    function_body.add(code_function_callt{
```
Hm, I thought you'd have to pass more than one parameter here sometimes?

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -249,7 +249,21 @@ code_blockt memory_snapshot_harness_generatort::add_assignments_to_globals(
+  code_blockt code{};
+
+  // add initialization for existing globals
+  for(auto pair : goto_model.symbol_table)
```
`const &`

xbauch replied:
Added.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c7c9708e41a864a73b8c1c4c2a2b93486cec270b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/126018554

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -249,7 +249,21 @@ code_blockt memory_snapshot_harness_generatort::add_assignments_to_globals(
+  {
+    if(
+      id2string(pair.first).find(CPROVER_PREFIX) != 0 &&
+      pair.second.is_static_lifetime)
```
See the function call harness, you'll also want 
```
&& symbol.is_lvalue 
&& symbol.type.id() != ID_code 
&& !has_prefix(id2string(symbol.name), CPROVER_PREFIX)
```

because we don't want to initialise function symbols, non-lvalues or cprover internal variables (we should probably factor out this filtering stuff into a separate utility)

xbauch replied:
Extracted and moved to `recursive_initializationt`.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -79,54 +81,172 @@ recursive_initializationt::recursive_initializationt(
+      "min_depth",
+      from_integer(
+        initialization_config.min_null_tree_depth,
+        signed_int_type())))
```
it's kind of irrelevant but I'd have expected these to be `size_t`

xbauch replied:
The only reason I went for `int` is that it looks better in the dumped code.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8cb2f6ed82a64f24075739ad4010e1f9437dfc73).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/126026333

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -79,54 +81,172 @@ recursive_initializationt::recursive_initializationt(
-    initialize_struct_tag(lhs, depth, known_tags, body);
-  }
-  else if(type.id() == ID_pointer)
+  const irep_idt &fun_name = build_constructor(lhs);
```
Surely `build_constructor` only needs `lhs.type()`?

xbauch replied:
We also need the name of the symbol (if `lhs` is a symbol) to get the `associated_size_variable` etc.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -79,54 +81,172 @@ recursive_initializationt::recursive_initializationt(
-    if(lhs.id() == ID_symbol)
+    const irep_idt &lhs_name = to_symbol_expr(lhs).get_identifier();
+    // skip initialisation of max/min depth globals
+    if(lhs_name == max_depth_var_name || lhs_name == min_depth_var_name)
```
I think it'd be cleaner to remove these from the global set in the first place - IIRC at least the function-harness automatically strips all variables with a `__CPROVER_` prefix from the list of variables to nondet (all harnesses should do this, may be worth extracting into separate functionality). All our global (maybe even local, given that cbmc doesn't really care all that much about scope) variables should have the `__CPROVER_` prefix attached to them anyway (that way we can avoid accidentally overriding variables that we add in here without having to remember to block each of them individually here)

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -79,54 +81,172 @@ recursive_initializationt::recursive_initializationt(
+          goto_model.symbol_table.lookup_ref(size_var.value()).symbol_expr();
+        body.add(code_function_callt{
+          fun_symbol.symbol_expr(),
+          {depth, address_of_exprt{lhs}, address_of_exprt{size_symbol}}});
```
That's one way to do it, but I don't like having the replicated logic between the constructor and the call; I'd prefer if the dynamic array constructor just always took a size pointer and just set it to null if it's not needed (kind of mirroring similar functions in the real world)

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -426,49 +332,421 @@ std::string recursive_initialization_configt::to_string() const
+{
+  symbolt &fresh_symbol = get_fresh_aux_symbol(
+    signed_int_type(),
+    "__goto_harness",
```
Probably makes more sense to use `CPROVER_PREFIX`, same in following

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -426,49 +332,421 @@ std::string recursive_initialization_configt::to_string() const
+
+symbol_exprt recursive_initializationt::get_free_function()
+{
+  auto free_sym = goto_model.symbol_table.lookup("free");
```
FYI we had some talk at some point about using CBMCs internal allocate/deallocate instructions for this, but I didn't have much luck for them (no need for changes, just a remark)

hannes-steffenhagen-diffblue reviewed (approved):
Mostly looks fine to me, but I think the decision of what things to initialise or not should be made outside of `recursive_initialisation`

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 099cf1208958436f0353c9c485cbc965077d10d6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128903045

NlightNFotis commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -428,6 +428,9 @@ std::string recursive_initialization_configt::to_string() const
 recursive_initializationt::array_convertert
 recursive_initializationt::default_array_member_initialization()
+irep_idt recursive_initializationt::get_fresh_global_name(
+  const std::string &symbol_name,
```
`get_fresh_global_name` and `get_fresh_global_symexpr` look very similar. Are we sure the bulk of the work performed in one function, and then have the helpers call into it?

NlightNFotis commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -121,7 +121,77 @@ void recursive_initializationt::initialize(
+  symbolt *mutable_symbol = symbol_table.get_writeable(function_symbol.name);
+
+  // the body is specific for each type of expression
+  mutable_symbol->value = build_constructor_body(
```
slightly weird that `build_constructor_body` is being referenced here but added in the next commit and there's a prototype added for it as well in `recursive_initialization.h`

xbauch replied:
I wasn't sure how to break the PR into individual commits so I sometimes formed a commit around a body of a function but only declared the functions called inside that body.

NlightNFotis commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -122,6 +122,40 @@ void recursive_initializationt::initialize(
+  {
+    if(lhs_name.has_value())
+    {
+      if(should_be_treated_as_cstring(*lhs_name) && type == char_type())
```
Just a question, since I'm not sure why it's happening: 

Why in other places we compare the `type.id()`, like for instance in line 137 `type.id() == ID_pointer` and here we compare the type itself with the constructor result? Are we sure equality is working as it should? If yes, should all of these be harmonised?

xbauch replied:
That's just my being lazy and not finding out what type-id `char` should have.

NlightNFotis reviewed (commented):
Looks good to me overall, just some minor comments.

petr-bauch committed:
Convert the whole model rather than a single function

because we also want to convert the function that will be added during
initialisation.

petr-bauch committed:
Clean-up harnesses in testing symbol init-ability

also add mode to memory-snapshot.

petr-bauch committed:
Helper functions to simplify manipulating symbols

and storing the in the symbol table (and one for `free`).

petr-bauch committed:
Rewrite the recursive constructors

Previously all recursive initialisation took place in a single function. Now
each type has it's own constructor function, which are called recursively. The
results (variables to be initialised) are passed by pointer.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9ca0824bb380132818fce832ca4e06dc733b2d5c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/130278481
